lib/Dialect/Seq/Transforms/RegOfVecToMem.cpp

```diff
@@ -0,0 +1,245 @@
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Seq/SeqPasses.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "reg-of-vec-to-mem"
+
+using namespace circt;
+using namespace seq;
+using namespace hw;
+
+namespace circt {
+namespace seq {
+#define GEN_PASS_DEF_REGOFVECTOMEM
+#include "circt/Dialect/Seq/SeqPasses.h.inc"
+} // namespace seq
+} // namespace circt
+
+namespace {
+
+struct MemoryPattern {
+  FirRegOp memReg;               // The register array representing memory
+  FirRegOp outputReg;            // Optional output register
+  Value clock;                   // Clock signal
+  Value readAddr;                // Read address
+  Value writeAddr;               // Write address
+  Value writeData;               // Write data
+  Value writeEnable;             // Write enable
+  Value readEnable;              // Read enable (optional)
+  comb::MuxOp writeMux;          // Mux selecting between old/new memory state
+  comb::MuxOp readMux;           // Mux for read data
+  hw::ArrayGetOp readAccess;     // Array read operation
+  hw::ArrayInjectOp writeAccess; // Array write operation
+};
+
+class RegOfVecToMemPass : public impl::RegOfVecToMemBase<RegOfVecToMemPass> {
+public:
+  void runOnOperation() override;
+
+private:
+  bool analyzeMemoryPattern(FirRegOp reg, MemoryPattern &pattern);
+  bool createFirMemory(MemoryPattern &pattern);
+  bool isArrayType(Type type);
+  std::optional<std::pair<uint64_t, uint64_t>> getArrayDimensions(Type type);
+
+  SmallVector<Operation *> opsToErase;
+};
+
+} // end anonymous namespace
+
+bool RegOfVecToMemPass::isArrayType(Type type) {
+  return isa<hw::ArrayType, hw::UnpackedArrayType>(type);
+}
+
+std::optional<std::pair<uint64_t, uint64_t>>
+RegOfVecToMemPass::getArrayDimensions(Type type) {
+  if (auto arrayType = dyn_cast<hw::ArrayType>(type)) {
+    auto elemType = arrayType.getElementType();
+    if (auto intType = dyn_cast<IntegerType>(elemType)) {
+      return std::make_pair(arrayType.getNumElements(), intType.getWidth());
+    }
+  }
+  return std::nullopt;
+}
+
+bool RegOfVecToMemPass::analyzeMemoryPattern(FirRegOp reg,
+                                             MemoryPattern &pattern) {
+  LLVM_DEBUG(llvm::dbgs() << "Analyzing register: " << reg << "\n");
+
+  // Check if register has array type
+  if (!isArrayType(reg.getType()))
+    return false;
+
+  pattern.memReg = reg;
+  pattern.clock = reg.getClk();
+
+  // Find the mux that drives this register
+  auto nextValue = reg.getNext();
+  auto mux = nextValue.getDefiningOp<comb::MuxOp>();
+  if (!mux)
+    return false;
+
+  LLVM_DEBUG(llvm::dbgs() << "  Found driving mux: " << mux << "\n");
+  pattern.writeMux = mux;
```

Comment:
Since the mux gets deleted later, it's probably necessary to check that the mux result isn't used anywhere outside this specific memory pattern. Might be worth adding a test for this.